libcrux-ml-dsa/src/key_centric_apis.rs

```diff
@@ -0,0 +1,640 @@
+                    payload,
+                    context,
+                    randomness.declassify(),
+                    signature,
+                )
+            }
+            /// Generate a HashML-DSA Signature, with a SHAKE128 pre-hashing
+            ///
+            /// The parameter `context` is used for domain separation
+            /// and is a byte string of length at most 255 bytes. It
+            /// may also be empty.
+            pub fn sign_pre_hashed_shake128(
+                key: &[U8; Self::SIGNING_KEY_LEN],
+                payload: &[u8],
+                signature: &mut [u8; Self::SIGNATURE_LEN],
+                context: &[u8],
+                randomness: [U8; 32],
+            ) -> Result<(), crate::SigningError> {
+                // TODO: use mut version
+                let mut pre_hash_buffer = [0; 32];
+                let signature_out =
+                    crate::ml_dsa_generic::multiplexing::$module::sign_pre_hashed_shake128(
+                        key.declassify_ref().try_into().unwrap(),
+                        payload,
+                        context,
+                        &mut pre_hash_buffer,
+                        randomness.declassify(),
+                    )?;
+
+                signature.copy_from_slice(signature_out.as_slice());
+
+                Ok(())
+            }
+
+            /// Verify an ML-DSA Signature
+            ///
+            /// The parameter `context` is used for domain separation
+            /// and is a byte string of length at most 255 bytes. It
+            /// may also be empty.
+            pub fn verify(
+                key: &[u8; Self::VERIFICATION_KEY_LEN],
+                payload: &[u8],
+                signature: &[u8; Self::SIGNATURE_LEN],
+                context: &[u8],
+            ) -> Result<(), crate::VerificationError> {
+                crate::ml_dsa_generic::multiplexing::$module::verify(
+                    key, payload, context, signature,
+                )
+            }
+
+            /// Verify an ML-DSA Signature, with a SHAKE128 pre-hashing
+            ///
+            /// The parameter `context` is used for domain separation
+            /// and is a byte string of length at most 255 bytes. It
+            /// may also be empty.
+            pub fn verify_pre_hashed_shake128(
+                key: &[u8; Self::VERIFICATION_KEY_LEN],
+                payload: &[u8],
+                signature: &[u8; Self::SIGNATURE_LEN],
+                context: &[u8],
+            ) -> Result<(), crate::VerificationError> {
+                let mut pre_hash_buffer = [0; 32];
+                crate::ml_dsa_generic::multiplexing::$module::verify_pre_hashed_shake128(
+                    key,
+                    payload,
+                    context,
+                    &mut pre_hash_buffer,
+                    signature,
+                )
+            }
+            /// Generate an ML-DSA Key Pair
+            pub fn keygen_derand(
```

Comment:
Again, I don't think we need the name to be keygen_derand if we only have the derand version. But I am not sure we need a rand version here since we already have KeyGen::generate.